### DISCUSSION TITLE
Best option for deploying in concurrent environment

### DISCUSSION THREAD

**Discussion by gthb619** [lancedb/lancedb#1843] [Q&A]
We use lance-db as a python lib, embedded within a streamlit app. The application is a chatbot answering questions following the RAG pattern. Retrieval is done by using lance-db and generation by invoking GPT. The preprocessed version of the documents used for retrieval is provided once to lance-db, which builds its tables and indices and is, then, ready to receive search requests and return chunks. Lance-db's tables and indices are not modified while answering to user questions. The building of lance-db's tables and indices takes less than a minute.

We are porting the app to cloud (Azure) and making it horizontally scalable. So many streamlit instances will need to access lance-db functionalities simultaneously for retrieval.

Which are the options? Should each streamlit instance access its own copy of lance-db, with private tables and indices? Or should we restructure the app so that there is only one instance of lance-db, wrapped by a REST API that gets accessed by many streamlit instances?

**Comment by QianZhu** [comment]
There could be concurrent queries to the same LanceDB table. So setting up one instance of LanceDB should be enough.
